sale_coupon_advanced_option/tests/test_first_so_program_only.py

```diff
@@ -0,0 +1,255 @@
+from odoo.exceptions import UserError
+
+from odoo.addons.sale_coupon.tests.common import TestSaleCouponCommon
+
+
+class TestProgramForFirstSaleOrder(TestSaleCouponCommon):
+    def setUp(self):
+        super(TestProgramForFirstSaleOrder, self).setUp()
+
+        self.env["sale.coupon.program"].search([]).write({"active": False})
+
+        self.product_A = self.env["product.product"].create(
+            {"name": "Product A", "list_price": 60, "sale_ok": True}
+        )
+
+        self.program1 = self.env["sale.coupon.program"].create(
```

Comment:
I am afraid test case is not working as you expected, with "discount_type": "percentage"  they have been filtered to have only the best one 
https://github.com/odoo/odoo/blob/114b9d2c7b858b9cc8bc8d82239f6fe53290d58f/addons/sale_coupon/models/sale_coupon_program.py#L150 
that is why I had different types of programs in my test case to be sure they all have been applied to my order
https://github.com/camptocamp/sale-workflow/blob/5e55e1686fb8579a71ce905dbd0b31c12d979083/sale_coupon_advanced/tests/test_sale_coupon_computation.py#L23
and in one test case you introducing previous behavior than a new one